### ISSUE TITLE
Diferent copy-files by platform

### ISSUE THREAD

**Issue by debelop13** [felixrieseberg/ember-electron#123]
Hello!

Can be possible to add different files (depending on platform) in packaging?

Thanks.

**Comment by danidr** [comment]
+1. I would also love to know this, since I am including a quite big (~500MB per architecture) external executable and including all platforms in all packages would make for a huge application.

**Comment by debelop13** [comment]
Hi! 
Finally I achieved this using --ignore like in this case:

ember electron:package --platform=darwin --ignore=folderToIgnore

ember electron:package --platform=win32 --ignore=folderToIgnore

**Comment by felixrieseberg** [comment]
Seems like a solid idea. If `copy-files` was also a command line parameter, your use case would be covered, right?

Would you be interested in making a PR?

**Comment by danidr** [comment]
Well, seeing that this issue will come up quite often, wouldn't it be better to create a sort of way to filter the `copy-files` options?
This could be extended to any type of setting, for example by creating a sub-category in the package.json for every platform (which could of course be optional, the property id could be set to 'all' and work for any type of platform). Like:
`
'win32': { options... },
'darwin': { options... }
`
At this point though, it would make more sense to create a config file specifically for ember-electron.

**Comment by anulman** [comment]
Hm. What if we relied on the fs directly?

#129 adds the notion of the `ember-electron` dir (since electron.js can't be at the same level as package.json). What if we concat'ed trees, broccoli style?

- `ember-electron/public/:platform` tree, merged into
- `ember-electron/public`; tree, merged into
-  `dist/public`

**Comment by bendemboski** [comment]
@anulman I like the idea of using the fs, but I think the platform sub-directory of `public` might be confusing. What do you think about making the platform public directories siblings of the "base" public directory:

- `ember-electron/public-:platform` tree, merged into
- `ember-electron/public`; tree, merged into
- `dist/public`

**Comment by anulman** [comment]
@bendemboski I've really liked your dash idea, but on second thought want to propose one more amendment: what about `-:platform` subdirs instead?

I think this better maps to a lot of Ember fs styleguides I've seen, and has the added benefits of being an easily globbable rule (`!.../-*/**/*`) + puts these special dirs up at the top of alphabetical lists.

To recap, the structure I'm proposing is:

- `ember-electron/public/-:platform/**/*`; merges into
- `ember-electron/public/**/*`, merges into
- `$packagedAppFiles/dist/public`

**Comment by bendemboski** [comment]
@anulman my main reason for proposing dashes was to avoid subdirectory nesting because that seems more confusing and potentially harder to manage to me. With nesting, the `public` directory contains a mix of files and directories that get copied as-is, and special directories that are handled specially.

I can point to the public directory and know it will get copied exactly as-is (and then possibly have some items overwritten with platform-specific files), and not have to worry about some directories being filtered out.

The logic for copy is simple (thinking in shell-ish commands):

```
$ cp -r public/* $builddir/assets/
$ cp -rf public-$platform/* $builddir/assets/
```

as opposed to
```
$ cp -r public/* $builddir/assets/
$ rm -rf $builddir/assets/-*
$ cp -rf public/-$platform/* $builddir/assets/
```

Also, corresponding files are at the same directory tree depth, meaning that `/public-win32/foo.jpg` overwrites `/public/foo.jpg`, both of which are at the same depth in the directory tree, rather than `/public/-win32/foo.jpg` overwriting `/public/foo.jpg`, which are at different depths.

Generally it's just easier for me to visualize a directory being copied on top of/merged with its sibling directory, rather than a directory being copied on top of/merged with its parent directory, except with a few special directories removed/ignored.

Anyway, I think it will work either way, so I won't complain if you disagree and want to go with nesting.

**Comment by pichfl** [comment]
We recently had a project with similar problems and I want to agree with @bendemboski on not nesting the directories. Makes things a lot easier.

**Comment by jacobq** [comment]
So would `electron.js` still go in `ember-electron/` or `ember-electron/public/`?

**Comment by pichfl** [comment]
`ember-electron/` I suppose. There should be no reason to maintain multiple versions of `electron.js` when you can handle platforms directly in code. 🙂

**Comment by bendemboski** [comment]
@pichfl @jacobq @anulman @felixrieseberg (anyone else interested), I think we need to have a slightly broader discussion of file/project structure. As I noted in #170, the fix for this issue kinda forces this discussion.

We used to have two basic ways of building the app:

1. When running `ember electron` we would build the Ember app into `dist/` and then run "in place" meaning `electron-forge` would look at the project's `package.json` whose main entry would point to `electron.js`, which would be run straight out of the source tree (`<project root>/ember-electron/electron.js`), so the Ember/`ember-electron` *was* the project that `electron-forge would operate on`.
2. When running `ember electron:test`, `ember electron:make` or `ember electron:package`, we would build the Ember app, and then assemble the full electron app in a temp directory, copying the files in `package.json`'s `copy-files` entry (which unfortunately happened in different places in code when running `ember electron:test` vs. the other two commands).

(1) is no longer possible once this issue is fixed, because now `ember-electron/public-<platform>` gets merged with `ember-electron/public`, so `electron.js` can no longer use require paths that are always valid in the source tree, because it obviously wouldn't work in a packaged app to require `public-<platform>/foo.js`.

So, we have to embrace the idea of always "building" the electron app (copying files somewhere and merging the public directories) and running against that built output, rather than running against the source tree, for all of our commands.

I think this is also good news, because now the structure of our packaged app is no longer constrained by the structure of our source tree. In particular, you can see in #170, we no longer need to modify the projects `package.json` to set `main` to point to `electron.js` because we can just do that during the build. But more generally, since we're using broccoli to build the electron app now, and makes it easy to move things around with ease, I think it's worth talking about how we actually want the source tree and running project to be structured.

So, right now the directory structure looks like this (ignoring files that aren't relevant here):

```
├── package.json
├── .compilerc
├── electron
│   ├── electron.js
│   ├── public
│       └── <files to copy>
│   ├── public-darwin
│       └── <files to copy for darwin>
│   ├── public-linux
│       └── <files to copy for linux>
│   ├── public-win32
│       └── <files to copy forwindow>
└── app
    └── <ember app source>
```

and the build output that I currently create in #170 is:

```
├── package.json
├── .compilerc
├── electron
│   ├── electron.js
│   └── <files copied/merged from public and public-<platform>>
└── ember
    └── <ember build output>
```

(with `package.json` modified to have a `main` entry pointing to `electron/electron.js`), and if we are running `ember electron:package` or `ember electron:make` there is a `node_modules` at the root of the tree as well containing all the production dependencies.

So what do you all think of that output directory structure? I kinda like keeping the electron app in its own sub-directory, and just having project/`electron-forge`-related files (`package.json` and `.compilerc`) in the root of the project. But here's another layout that we could have that would also make sense:

```
├── package.json
├── .compilerc
├── electron.js
└── <file copied/merged from public and public-<platform>>
└── <file copied/merged from public and public-<platform>>
└── ...
└── ember
    └── <ember build output>
```

That basically makes the project and the electron app no longer distinct from each other, and the Ember app is just in a subdirectory. Other than us looking at it and deciding which is the most intellectually satisfying, is there a reason it would make more/less sense to put `electron.js` at the root of the project? I can't think of one -- plenty of projects have their `main` script at the root, and plenty of them have it in a `src` or `lib` directory, so I'm not sure there's a clear convention here.

As I said, I think I prefer it as-is, but wanted to solicit some feedback.

I also think that in the source tree we should move `electron.js` into the `public` directory. I agree with @pichfl (https://github.com/felixrieseberg/ember-electron/issues/123#issuecomment-285101823) that there isn't really a need for platform-specific `electron.js`'s, but who knows when some user might want to for some reason, and why stop them? Also, it makes sense from a project structure standpoint, because right now in the source tree if `electron.js` wants to import a file that is at `public/foo.js`, it would need to `require('./foo.js')` (because they end up in the same folder in the output tree). That's fine because we never run `electron.js` from the source tree, but I think would be confusing to people reading the code. In summary, my argument is that `electron.js` can be treated as just another file at the root of the electron app (that happens to be special in that it's the main module), so let's treat it that way in the build.

So, to distill this down to the two questions on which I am requesting feedback (sorry, I get kinda wordy sometimes):

1. Do folks like the current build output directory structure, or do you prefer the other one I proposed, or another one I didn't think of?
1. Does anybody object to putting `electron.js` in the `public/` directory in the source tree?

**Comment by felixrieseberg** [comment]
> So, we have to embrace the idea of always "building" the electron app (copying files somewhere and merging the public directories) and running against that built output, rather than running against the source tree, for all of our commands.

This terrifies me, but I understand why it'd be in our (current) interest to do that. With that in mind, I prefer the current directory structure you proposed in the PR (https://github.com/felixrieseberg/ember-electron/pull/170) - I believe it's easier to reason about, given that it's closer to the actual input.

There's just one thing I'm against: Platform-specific `electron.js` files. That seems like a bad practice:tm: to me, so our default should be to just keep it at the top. Platform-specific static files (you'd usually expect in a public folder) are cool, platform-specific code should just be gated in-code - to match virtually all major Electron apps out there. Basically, we wouldn't make an `ember-chrome` or `ember-firefox` folder, so we shouldn't try to do the same thing for systems :wink:

tl;dr:
:one: I prefer the structure you introduced in #170
:two: Kind of object, let's leave it at top of `electron`

**Comment by jacobq** [comment]
DISCLAIMER: I'm still learning about how all the pieces get put together to make a full ember + electron app. That aside, here're my thoughts:

1. I like the first layout (like #170) better because it feels akin to a natural front-end (ember / renderer) / back-end (electron / main process) separation. Relative paths often throw me for a loop with projects where the source gets transformed, but as long as its clear what the path should be and as long as `ember electron` and `ember electron:[package|make]` follow the same convention then it shouldn't be a big deal.

2. Though I raised the question earlier I hope I didn't suggest that `electron.js` should be moved into `public`. Although that would eliminate a small amount of logic in the build process (wouldn't need to treat it specially) I think it's nice for users to start with an empty space (i.e. `public`) as otherwise the answer to the question, "Is this where I'm supposed to put stuff?" isn't as immediately clear. 

tl;dr: I agree with @felixrieseberg

**Comment by bendemboski** [comment]
@felixrieseberg I get what you're saying, but that's not how it's currently structured. Other than `electron.js` which is special, if you want any node code in your source tree to get propagated to the output, it has to be in a `public/` directory. That's because the old mechanism for both code and resources was `copy-files`, and all we did was convert it to the `public/` directories, so now those are the mechanism for both code and resources.

So, I think we're kinda stuck with what we have unless we want to separate out code from resources in the source tree. I don't think that's a bad idea -- similar to how Ember does it with the `app` and `public` directories. We could create a `lib` directory in `ember-electron` as a peer of the `public` and `public-<platform>` directories like Ember does. We would put `electron.js` in that directory and tell people to put other in-repo code there as well.

Then we would copy the `lib` directory to a `lib` directory in the output, and the same with the `public` directory.

So the source tree would look like:

```
├── package.json
├── ember-electron
│   ├── .compilerc
│   └── lib
│       ├── electron.js
│       └── <code to copy>
│   ├── public
│       └── <resource files to copy>
│   ├── public-darwin
│       └── <resource files to copy for darwin>
│   ├── public-linux
│       └── <resource files to copy for linux>
│   ├── public-win32
│       └── <resource files to copy for windows>
└── app
    └── <ember app source>
```

and the build output could look like:

```
├── package.json
├── .compilerc
├── lib
│   └── <contents of lib including electron.js>
├── public
│   ├── <file copied/merged from public and public-<platform>>
│   ├── <file copied/merged from public and public-<platform>>
│   └── ...
└── ember
    └── <ember build output>
```

Although if we're doing it that way, perhaps we should call `public` something different? Like `resources` or `assets` (similar to Ember) or just `res` or something? Thoughts?

**Comment by jacobq** [comment]
I would appreciate it if we avoided using names that are used elsewhere unless they have some kind of prefix or other differentiating feature since having multiple different things with the same name can be confusing. e.g. If someone says, "I have file X in public. Why doesn't my app see it?" then one first has to clarify, "Is it public or ember-electron/public?", etc.

**Comment by bendemboski** [comment]
@jacobq yeah, that makes sense. So I guess my updated updated updated proposal is:

Source tree

```
├── package.json
├── ember-electron
│   ├── .compilerc
│   └── lib
│       ├── electron.js
│       └── <code to copy>
│   ├── resources
│       └── <resource files to copy>
│   ├── resources-darwin
│       └── <resource files to copy for darwin>
│   ├── resources-linux
│       └── <resource files to copy for linux>
│   ├── resources-win32
│       └── <resource files to copy for windows>
└── app
    └── <ember app source>
```

and the build output could look like:

```
├── package.json
├── .compilerc
├── lib
│   └── <contents of lib including electron.js>
├── resources
│   ├── <file copied/merged from resources and resources-<platform>>
│   ├── <file copied/merged from resources and resources-<platform>>
│   └── ...
└── ember
    └── <ember build output>
```

**Comment by anulman** [comment]
+1 to this; I'm also thinking we should add `forge.json` to the source tree. https://github.com/felixrieseberg/ember-electron/pull/170/files#r105287096

**Comment by pichfl** [comment]
Can we put the out directory to `/ember-electron-out`?  I think it's a little weird to end up with the output in a resource dir.

**Comment by jacobq** [comment]
In that case, why not use the previous default of `electron-builds`?

**Comment by anulman** [comment]
I wanted to match forge standards (they use `out`). `src/out`?

**Comment by jacobq** [comment]
I like `out` at the top level of the project. Is that bad form?

**Comment by pichfl** [comment]
How about something like this? 

```
├── package.json
├── electron
|   ├── .compilerc
|   ├── lib
|   |   ├── electron.js
|   |   └── <code to copy>
│   ├── public
│   |   └── <resource files to copy>
│   ├── public-darwin
│   |   └── <resource files to copy for darwin>
│   ├── public-linux
│   |   └── <resource files to copy for linux>
│   ├── public-win32
│   |   └── <resource files to copy for windows>
│   └── out
|       └── <build output>
└── app
    └── <ember app source>
```

Maybe putting /electron/out to /dist-electron might be interesting as well, trying to align to the way an ember application is built and the output is found in /dist

**Comment by anulman** [comment]
Good thought @pichfl, though I disagree per @jacobq's earlier note of confusing ember public with electron public, and also because `dist` is meant as its own distributable (vs. this dir which wraps several possible build outputs).

**Comment by pichfl** [comment]
okay. so would `/electron/out` or `/electron-out` be more acceptable? I hope I followed this correctly.

**Comment by bendemboski** [comment]
I definitely agree with moving the package output out of `ember-electron`, but I agree with @anulman that calling it something like `dist` could be confusing. I'd vote for `electron-out`.

@anulman so your proposal would be to essentially make the `ember-electron` source directory look just like the built output, except it would of course be missing the `ember` directory, and would have the unmerged `resources-<platform>` directories? I guess that makes sense -- we would be saying that we just give you a mini-project to populate with the special feature of platform-specific `resources` directories, and we'll take care of adding the ember build and package.json for you.

I suppose we could even support you putting a custom package.json in the `ember-electron` directory rather than using the project's top-level package.json (and maybe in the broccoli pipeline ensure that it has a few minimal dependencies, like `electron-compile`).

Also, I agree with renaming `electron.js` to `main.js`.

Unless anybody speaks up objecting, I'll make these changes. So we'll have:

Source tree

```
├── package.json
├── ember-electron
│   ├── .compilerc
│   ├── index.js (formerly electron.js)
│   ├── other-source-file.js
│   ├── other-dir
│   |   ├── some-other-code.js
│   |   └── some-other-thingy.json
│   ├── resources
│   |   └── <resource files to copy>
│   ├── resources-darwin
│   |   └── <resource files to copy for darwin>
│   ├── resources-linux
│   |   └── <resource files to copy for linux>
│   └── resources-win32
│       └── <resource files to copy for windows>
├── app
│   └── <ember app source>
├── dist
│   └── <ember build output>
└── electron-out
    ├── my-app-darwin-x64
    │   └── etc...
    ├── my-app-win32-x64
        └── etc...
    └── make
        └── etc...
```

Output tree

```
├── package.json
├── .compilerc
├── index.js
├── other-source-file.js
├── other-dir
│   ├── some-other-code.js
│   └── some-other-thingy.json
├── resources
│   ├── <file copied/merged from resources and resources-<platform>>
│   ├── <file copied/merged from resources and resources-<platform>>
│   └── ...
└── ember
    └── <ember build output>
```

**Comment by bendemboski** [comment]
Crap. `electron-forge` won't let our main module be at the root of the project. Updated previous comment to put it under `lib/`.

**Comment by anulman** [comment]
Lol, derp. Forgot that restriction.

This looks great. Thanks so much Ben!

**Comment by jacobq** [comment]
I like what you're doing here, but I must've missed the discussion about changing the name of `electron.js` to `index.js`. Is the intention to be more similar to other conventions (main, index, entry, start, etc.), a restriction to make it easier to interface with other things, or something else? I trust your instincts more than mine, so I'm not opposed to it, but I would like to understand the motivation. To me it feels like moving from a more informative name to a more generic one.

**Comment by bendemboski** [comment]
The idea is that we're making the ember-electron directory look almost
exactly like its own node package, and it's very common for the main module
of a package to be named index.js. So it should look familiar through that
context, and since it's in a directory called "ember-electron" in the
source tree I don't think it will cause any confusion.

B

**Comment by felixrieseberg** [comment]
```
├── package.json
├── ember-electron
│   ├── .compilerc
│   ├── lib
│   │   └── index.js (formerly electron.js)
│   ├── other-source-file.js
│   ├── other-dir
│   |   ├── some-other-code.js
│   |   └── some-other-thingy.json
│   ├── resources
│   |   └── <resource files to copy>
│   ├── resources-darwin
│   |   └── <resource files to copy for darwin>
│   ├── resources-linux
│   |   └── <resource files to copy for linux>
│   └── resources-win32
│       └── <resource files to copy for windows>
├── app
│   └── <ember app source>
├── dist
│   └── <ember build output>
└── electron-out
    ├── my-app-darwin-x64
    │   └── etc...
    ├── my-app-win32-x64
        └── etc...
    └── make
        └── etc...
```

🙌

**Comment by jacobq** [comment]
I think this can be closed now that #170 has landed, right?
